File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;

        ListNode counter = head;
        int totalNodes = 0;
        while (counter != null)    {
            totalNodes++;
            counter = counter.next;
        }

        ListNode prev = dummy;
        ListNode curr = prev.next;
        while (totalNodes >= k) {
            ListNode groupStart = curr;
            ListNode groupNext = null;

            for (int i = 0; i < k; i++) {
                ListNode temp = curr.next;
                curr.next = groupNext;
                groupNext = curr;
                curr = temp;
            }

            prev.next = groupNext;
            groupStart.next = curr;

            prev = groupStart;
            totalNodes -= k; 
        }

        return dummy.next;
    }
}

File: Data Structures & Algorithms/reverse-a-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # 'prev' will point to the previous node (initially None)
        prev = None

        # 'curr' starts at the head of the list
        curr = head

        # Traverse the list
        while curr:
            # Store next node before breaking the link
            nxt = curr.next

            # Reverse the current node's pointer
            curr.next = prev

            # Move 'prev' and 'curr' one step forward
            prev = curr
            curr = nxt

        # At the end, 'prev' will be the new head of the reversed list
        return prev